tests/testthat/test-interface.R

```diff
@@ -254,3 +254,42 @@ test_that("Can get parameters from generators", {
   expect_equal(coef(mod), coef(res))
   expect_equal(coef(mod), mod$param())
 })
+
+
+test_that("can change data", {
+  res <- dust_example("walk")
+
+  obj <- res$new(list(sd = 1), 0L, 10L, seed = 1L)
+  y1 <- obj$run(1)
+
+  obj$set_data(list(sd = 2))
+  expect_equal(obj$state(), y1)
+  expect_equal(obj$step(), 1)
+  expect_equal(obj$data(), list(sd = 2))
+
+  y2 <- obj$run(2)
+
+  ## Then the comparison:
+  cmp <- dust_rng$new(seed = 1, 10)
+  expect_equal(cmp$rnorm(10, 0, 1), drop(y1))
+  expect_equal(y1 + cmp$rnorm(10, 0, 2), y2)
+})
+
+
+test_that("Validate changing data leaves particles in sensible state", {
+  res <- dust_example("variable")
+
+  obj <- res$new(list(len = 5, mean = 0, sd = 1), 0, 10, seed = 1L)
+  y1 <- obj$run(1)
+
+  expect_error(
+    obj$set_data(list(len = 6, mean = 10, sd = 10)),
+    paste("Tried to initialise a particle with a different size:",
+          "particle 1 had length 5 but data implies size 1"))
+  expect_identical(obj$state(), y1)
```

Comment:
Still not totally sure what 'length' is in the error here. Is it if you change some of the internal variable's dimensions?

Reply:
yeah, this is unclear. I've updated the docs